### ISSUE TITLE
withTestApplication sendPipeline interceptors do not work

### ISSUE THREAD

**Issue by lamba92** [ktorio/ktor#1557]
**Ktor Version and Engine Used (client or server and name)**
1.2.6, 1.3.0-rc2

**Describe the bug**
Any `sendPipeline` interceptor installed does not get called when testing using `withTestApplication`.

**To Reproduce**
Install a `sendPipeline` interceptor an try to use it with a `handleRequest`. The interception lambda will not be called.

**Expected behavior**
The interception lambda to be called.

You can check out the [test](https://github.com/lamba92/ktor-spa/blob/master/src/jvmTest/kotlin/com/github/lamba92/ktor/features/Tests.kt) code here in [this feature](https://github.com/lamba92/ktor-spa/blob/master/src/jvmMain/kotlin/com/github/lamba92/ktor/features/SinglePageApplication.kt) of mine where I'm having the issue. A run of the tests is also available [here](https://travis-ci.org/lamba92/ktor-spa/builds/635388649) on TravisCI and the corresponding [build scan](https://gradle.com/s/no4wmc5swmx26) here.

**Comment by cy6erGn0m** [comment]
Just added a test and it works well

```kotlin
@Test
    fun testSendPipeline(): Unit = withTestApplication {
        var processed = false

        application.sendPipeline.intercept(ApplicationSendPipeline.Before) {
            processed = true
        }

        application.intercept(ApplicationCallPipeline.Call) {
            call.respondText("OK")
        }

        assertEquals("OK", handleRequest(HttpMethod.Get, "/").response.content)
        assertTrue(processed)
    }
```

Test passes

**Comment by cy6erGn0m** [comment]
With the after phase works as well

**Comment by lamba92** [comment]
I have no idea why my tests do not pass then. The response is `null` and the `CallLogging` says the request is unhandled.

**Comment by cy6erGn0m** [comment]
Well, then the send pipeline is not executed because no response was sent

**Comment by lamba92** [comment]
Well yes, but wouldn't Ktor automatically respond with an empty body and 404 code?

**Comment by cy6erGn0m** [comment]
No, otherwise there will be no "unhandled"

**Comment by cy6erGn0m** [comment]
However, it's easy to add a fallback call interceptor that produces 404 if unhandled.

**Comment by cy6erGn0m** [comment]
Something like this

```kotlin
application.intercept(ApplicationCallPipeline.Fallback) {
    if (call.response.status() == null) {
        call.respond(HttpStatusCode.NotFound)
        finish()
    }
}
```

**Comment by lamba92** [comment]
May I ask if you are addressing this inconsistency?

**Comment by cy6erGn0m** [comment]
I'm almost ready to make that decision :)

**Comment by cy6erGn0m** [comment]
This couldn't be integrated before 1.4.0 because this is breaking change

**Comment by lamba92** [comment]
[Oh well...](https://www.youtube.com/watch?v=yJxCdh1Ps48) I'll just wait then! Thanks :)

**Comment by oleg-larshin** [comment]
Please check the <a href="https://youtrack.jetbrains.com/issue/KTOR-493">following ticket</a> on YouTrack for follow-ups to this issue. GitHub issues will be closed in the coming weeks.

**Comment by e5l** [comment]
Fixed in 1.4.0
